### PR TITLE
150 extend non native policies

### PR DESCRIPTION
**Pull request by CblPOK-git**

### CODE REVIEW

**Comment by Iluvmagick** [comment]
Unrelated, but: why do ordinary addition/multiplication/division depend on NonNativePolicyType? It doesn't seem to be used there at all.
I understand why would **non-native** variants require a non-native policy specification, but it being present as a template parameter for **native** ones confuses me.

**Comment by Iluvmagick** [comment]
`blueprint_algebra_fields_plonk_non_native_reduction_test` does not build for me.
```c++
In file included from libs/crypto3/libs/blueprint/test/algebra/fields/plonk/non_native/reduction.cpp:40:
libs/crypto3/libs/blueprint/include/nil/blueprint/components/algebra/fields/plonk/non_native/reduction.hpp:75:92: error: no type named 'non_native_var_type' in 'nil::blueprint::detail::basic_non_native_policy_field_type<typename crypto3::algebra::curves::pallas::base_field_type, typename crypto3::algebra::curves::ed25519::scalar_field_type>'
                    typename non_native_policy_type::template field<operating_field_type>::non_native_var_type output;
```
You need to update the type for <pallas::base, ed25519::scalar> too.

**Review by Iluvmagick — CHANGES_REQUESTED** [review]
Reduction test should be working and we shouldn't have to reimplement the same functions for each policy we create.

File: include/nil/blueprint/basic_non_native_policy.hpp [hunk 1]
```diff
@@ -65,6 +95,51 @@ namespace nil {
...
+                using native_field_type = typename crypto3::algebra::curves::pallas::base_field_type;
+                using var = crypto3::zk::snark::plonk_variable<native_field_type>;
+
+                typedef std::array<var, ratio> non_native_var_type;
+                typedef std::array<native_field_type::value_type, ratio> chopped_value_type;
+                
+                constexpr static const std::array<std::size_t, ratio> chunk_sizes = {254, 1};
+
+
+                static native_field_type::value_type get_i_th_chunk(non_native_field_type::value_type input,
```
[the remark this answers is not in the document]
Comment (CblPOK-git) [reply include/nil/blueprint/basic_non_native_policy.hpp:116]: Fixed reduction test, now works.
Copy-paste of function is done on purpose, we are going to re-write this thing https://github.com/NilFoundation/zkllvm-blueprint/issues/164 , it was fast way to implement needed feature. Not worth spending time optimizing the code that will be re-written soon.

**Comment by CblPOK-git** [comment]
> why do ordinary addition/multiplication/division depend on NonNativePolicyType

In the future policies will be used in native components too, just added and not implemented yet

**Review by nkaskov — APPROVED** [review]

**Review by nkaskov — APPROVED** [review]

**Review by Iluvmagick — APPROVED** [review]

**Pull request merged**